### PR TITLE
Support interfaces as translation context

### LINKED ISSUE #28: Narrow interfaces are not assignable to TranslationContext
`ctx.t` accepts `TranslationContext` that is defined as `Record<string, FluentValue>`
However, if you try to pass a narrower type (like `{a: string}`), you get an error

### PR DESCRIPTION
**Pull request by Loskir**
Fixes #28

### CODE REVIEW

**Review by KnorpelSenf — APPROVED** [review]
I'd like to have another review on this, but to me, it looks good

**Comment by KnorpelSenf** [comment]
ci is mean to you today

File: src/types.ts
Comment (EdJoPaTo) [inline src/types.ts]: Can we name it not something like Context? In relation with grammY it is confusing for users to have two contexts which are different things.

Fluent itself seems to call them ["placeables"](https://projectfluent.org/fluent/guide/placeables.html).

Comment (Loskir) [reply src/types.ts]: That's taken from https://deno.land/x/better_fluent@v1.0.0/mod.ts?s=TranslationContext

Comment (EdJoPaTo) [reply src/types.ts]: Also "Custom" is a bit… there only is one type for that, so its not a "custom" kind of it, its the default one.

Comment (EdJoPaTo) [reply src/types.ts]: > That's taken from https://deno.land/x/better_fluent@v1.0.0/mod.ts?s=TranslationContext

Yeah, I know, but we dont need to do something another library does especially when the project itself doesnt seem to use the word "context" at all :thinking: 
<https://duckduckgo.com/?q=site%3Aprojectfluent.org+context>

Comment (EdJoPaTo) [reply src/types.ts]: Placeables seem to be in a translation, variables are filled into these placeables. So placeable would be wrong too. But naming them variables might not be that helpful.

Comment (EdJoPaTo) [reply src/types.ts]: Dont get me wrong, I see where the name is coming from. But I think we can do better.

Comment (KnorpelSenf) [reply src/types.ts]: I do have to agree that the terms custom and context are usually combined when referring to something that is very different …

Comment (EdJoPaTo) [reply src/types.ts]: How about just naming it `TranslationVariables`?

Results basically in this type which seems easy to understand what it does and what its for:

```typescript
export type TranslationVariables<K extends string> = Record<K, FluentVariable>;
```

Bonus points for accepting `Readonly<TranslationVariables>` as function argument. 😎

Comment (KnorpelSenf) [reply src/types.ts]: I like that name! @Loskir?

Comment (Loskir) [reply src/types.ts]: Me too

Comment (Loskir) [reply src/types.ts]: Hmm, what's the point of Readonly?

Comment (Loskir) [reply src/types.ts]: @dcdunkan what do you think? it would be cool to add these changes to better_fluent itself. However, it's probably breaking

Comment (EdJoPaTo) [reply src/types.ts]: Readonly helps to ensure that the function does not have side effects on the input and does not change the input variable. (But its still only TypeScript, JavaScript has no types nor the readonly stuff and its not recursive. Not as good as I would like it to be but better than nothing)

Comment (dcdunkan) [reply src/types.ts]: > @dcdunkan what do you think? it would be cool to add these changes to better_fluent itself. However, it's probably breaking

I agree with it. But I'm away from my PC and I won't be able to work on it for a few months. I'd appreciate if someone could work on this and create a PR to better_fluent. I can review and merge the changes.

File: src/i18n.ts
Comment (ngdangtu-vn) [inline src/i18n.ts]: I would like use `ctx.translate` to define function and `ctx.t` as an alias. It will be more readable that way if we go like this.

```Typescript
ctx.translate = <K extends string>(
  key: string,
  translationContext?: TranslationVariables<K>,
): string => {
  return translate(key, {
    ...globalTranslationContext?.(ctx),
    ...translationContext,
  });
};
ctx.t = ctx.translate;
```

This way developers can easily understand what the function is about.

**Review by ngdangtu-vn — CHANGES_REQUESTED** [review]
Overall it looks good. I ran the test on my pc (linux) and it works fine. This update of interface and code refactor should work well.

**Comment by KnorpelSenf** [comment]
@Loskir are you still working on this?

**Comment by Loskir** [comment]
Should be ready to ship

**Review by KnorpelSenf — APPROVED** [review]
Looks good to me, thanks!

Didn't run it myself, though, so perhaps you wanna wait for someone else to review :)

**Review by EdJoPaTo — APPROVED** [review]

**Review by ngdangtu-vn — APPROVED** [review]

**Comment by KnorpelSenf** [comment]
Can you explain why it was closed?

**Comment by rojvv** [comment]
@KnorpelSenf:

image

**Comment by KnorpelSenf** [comment]
Transparency matters 🤓

Included in #35.

**Comment by dcdunkan** [comment]
> Can you explain why it was closed?

Ooh, sorry. I thought I wrote a comment about it, and I didn't 😬

**Pull request closed without merging**
